examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayer-JNI.cpp

```diff
@@ -0,0 +1,160 @@
+    {
+        ChipLogProgress(AppServer,
+                        "CastingPlayer-JNI::VerifyOrEstablishConnection() calling CastingPlayer::VerifyOrEstablishConnection() on "
+                        "Casting Player with device ID: %s",
+                        castingPlayer->GetId());
+        castingPlayer->VerifyOrEstablishConnection(callback, static_cast<unsigned long long int>(commissioningWindowTimeoutSec));
+    }
+    else
+    {
+        // Convert the EndpointFilter Java class to a C++ EndpointFilter
+        jclass endpointFilterJavaClass = env->GetObjectClass(desiredEndpointFilterJavaObject);
+        jfieldID vendorIdFieldId       = env->GetFieldID(endpointFilterJavaClass, "vendorId", "Ljava/lang/Integer;");
+        jfieldID productIdFieldId      = env->GetFieldID(endpointFilterJavaClass, "productId", "Ljava/lang/Integer;");
+        jobject vendorIdIntegerObject  = env->GetObjectField(desiredEndpointFilterJavaObject, vendorIdFieldId);
+        jobject productIdIntegerObject = env->GetObjectField(desiredEndpointFilterJavaObject, productIdFieldId);
+        // jfieldID requiredDeviceTypesFieldId = env->GetFieldID(endpointFilterJavaClass, "requiredDeviceTypes",
+        // "Ljava/util/List;");
+
+        matter::casting::core::EndpointFilter desiredEndpointFilter;
+        // Value of 0 means unspecified
+        desiredEndpointFilter.vendorId  = vendorIdIntegerObject != nullptr
+             ? static_cast<uint16_t>(env->CallIntMethod(
+                  vendorIdIntegerObject, env->GetMethodID(env->GetObjectClass(vendorIdIntegerObject), "intValue", "()I")))
+             : 0;
+        desiredEndpointFilter.productId = productIdIntegerObject != nullptr
+            ? static_cast<uint16_t>(env->CallIntMethod(
+                  productIdIntegerObject, env->GetMethodID(env->GetObjectClass(productIdIntegerObject), "intValue", "()I")))
+            : 0;
+        ChipLogProgress(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() desiredEndpointFilter.vendorId: %d",
+                        desiredEndpointFilter.vendorId);
+        ChipLogProgress(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() desiredEndpointFilter.productId: %d",
+                        desiredEndpointFilter.productId);
+        // TODO: In following PRs. Translate the Java requiredDeviceTypes list to a C++ requiredDeviceTypes vector. For now we're
+        // passing an empty list of.
+
+        ChipLogProgress(AppServer,
+                        "CastingPlayer-JNI::VerifyOrEstablishConnection() calling "
+                        "CastingPlayer::VerifyOrEstablishConnection() on Casting Player with device ID: %s",
+                        castingPlayer->GetId());
+        castingPlayer->VerifyOrEstablishConnection(callback, static_cast<unsigned long long int>(commissioningWindowTimeoutSec),
+                                                   desiredEndpointFilter);
+    }
+
+    return completableFutureObjGlobalRef;
```

Comment:
who will delete this completableFutureObjGlobalRef? potentially global ref leak? I have only found one env->DeleteGlobalRef(completableFutureObjGlobalRef); in this function at line 108?

if you release this in line 108, then what do you expect to return one released completableFutureObjGlobalRef?

Reply:
completableFutureObjGlobalRef gets released on line 108, when the callback defined (starting on line 74) gets called. This callback gets called in All cases, when castingPlayer->VerifyOrEstablishConnection(callback, ...) line 119 succeeds or fails. 
It is deleted after being used in CallBooleanMethod on lines 91 (success) or 106 (failure).

Reply:
I see, thanks